docs/wiki/EL/devp2p.md

```diff
@@ -2,6 +2,286 @@
+| `ip6`       | IPv6 address (16 bytes)                              |
+| `tcp6`      | IPv6 TCP port (big-endian integer)                   |
+| `udp6`      | IPv6 UDP port (big-endian integer)                   |
+
+*Note*: All keys except `id` are optional. 
+
+
+
+#### "v4" Identity Scheme
+
+The default identity scheme is `"v4"`, which is compatible with Node Discovery v4.
+
+more info about [Ethereum Node Records](https://github.com/ethereum/devp2p/blob/master/enr.md)
+
+### 2. Discovery Protocols (Discv4 & Discv5)
+
+Ethereum uses a **Node Discovery Protocol** to enable decentralized peer-to-peer networking. The protocol helps nodes locate and connect with each other efficiently. 
+Two versions exist:  
+
+- **Discv4** 
+- **Discv5**
+
+#### Discv4
+A structured, distributed system that allows Ethereum nodes to discover peers without central coordination.  
+
+- **Node Identities**  
+  - Each node is identified by a secp256k1 key pair.  
+  - The public key serves as the node’s unique identifier (Node ID).  
+  - Distance between nodes is computed using XOR of hashed public keys.  
+
+- **Node Records (ENR)**  
+  - Nodes store and share connection details using Ethereum Node Records (ENRs).  
+  - The "v4" identity scheme is used to verify node authenticity.  
+  - Peers can request a node’s latest ENR via an **ENRRequest** packet.  
+
+- **Kademlia Table**  
+  - Nodes maintain a **routing table** with 256 **k-buckets** (each holding up to 16 entries).  
+  - A bucket stores nodes within a specific distance range (e.g., `[2^i, 2^(i+1))`).  
+  - Nodes are sorted by last-seen time, ensuring stale nodes are replaced when the table is full.  
+
+- **Endpoint Verification (Proof-of-Participation)**  
+  - Prevents amplification attacks by verifying nodes before responding to queries.  
+  - A node is considered verified if it has sent a valid **Pong** response to a recent **Ping** request.  
+
+- **Recursive Lookup Algorithm**  
+  - Finds the `k` (typically 16) closest nodes to a target.  
+  - The search begins by querying a small, selected subset of the closest known nodes (`α`, often set to 3).  
+  - The lookup is **iterative**, querying new nodes found in previous steps until no closer nodes are discovered.  
+
+- **Wire Protocol & Message Types**  
+  - Messages are sent over **UDP** .  
+  - Each packet contains a header (`hash`, `signature`, `packet-type`) followed by encoded data.  
+  - Core message types:  
+    - **Ping (0x01):** Verify node availability.  
+    - **Pong (0x02):** Response to a Ping, proving reachability.  
+    - **FindNode (0x03):** Request nodes near a target ID.  
+    - **Neighbors (0x04):** Reply to FindNode with closest known peers.  
+    - **ENRRequest (0x05):** Request a node’s latest ENR.  
+    - **ENRResponse (0x06):** Provide an ENR in response to a request.  
+
+
+#### Discv5
+
+discv5 enables nodes to find and connect with peers without relying on centralized directories.
```

Comment:
Both Discv4 and Discv5 are decentralized, so maybe rephrasing to highlight Discv5's improvements over Discv4 in terms of service discovery and security.